hazelcast/src/main/java/com/hazelcast/jet/impl/processor/AsyncTransformUsingServiceUnorderedP.java

```diff
@@ -144,36 +168,87 @@ private boolean processItem(@Nonnull T item) {
             return true;
         }
         asyncOpsCounter++;
-        watermarkCounts.merge(lastReceivedWm, 1L, Long::sum);
-        Long lastWatermarkAtReceiveTime = lastReceivedWm;
+        for (int i = 0; i < wmKeys.length; i++) {
+            watermarkCounts[i].merge(lastReceivedWms[i], 1, Integer::sum);
+        }
+
+        long[] lastWatermarksAtReceiveTime = lastReceivedWms;
```

Comment:
The `lastWatermarksAtReceiveTime` variable is only used in the when-complete action of the future, and is needed as an old copy of `lastReceivedWms`. So it makes sense to declare it as close as possible...

Reply:
Yes but `lastWatermarksAtReceiveTime`, must be the watermarks used for incrementing watermark counts, because that will be used when decrementing when future completes. So let's say when the code was at line 174, if somebody else were to modify watermark counts, wrong copy will be sent to future. However this can never happen as Processor is single threaded. So this is nit-picking. I will mark this as resolved.